File: src/Skland/Http.hs
module Skland.Http
  ( getAuthorization
  , getCredential
  , getBindingList
  , signArknights
  , signEndfield
  )
where

import System.Exit (ExitCode (..))
import System.Process (CreateProcess, proc, readCreateProcessWithExitCode)

import Skland.Json
import Skland.Sign
import Skland.Types

userAgent :: String
userAgent = "Mozilla/5.0 SKLand/1.52.1"

baseHeaders :: String -> [(String, String)]
baseHeaders did =
  [ ("User-Agent", userAgent)
  , ("Accept-Encoding", "gzip")
  , ("Connection", "close")
  , ("X-Requested-With", "com.hypergryph.skland")
  , ("dId", did)
  ]

getAuthorization :: String -> UserConfig -> IO (Either String String)
getAuthorization did user = do
  let body =
        "{\"appCode\":\"4ca99fa6b56cc2ba\",\"token\":"
          <> compactString (userToken user)
          <> ",\"type\":0}"
  response <- curlPost "https://as.hypergryph.com/user/oauth2/v2/grant" (baseHeaders did) body
  pure (response >>= parseAuthorization)

getCredential :: String -> String -> IO (Either String Credential)
getCredential did code = do
  let body = "{\"code\":" <> compactString code <> ",\"kind\":1}"
  response <- curlPost "https://zonai.skland.com/web/v1/user/auth/generate_cred_by_code" (baseHeaders did) body
  pure (response >>= parseCredential)

getBindingList :: String -> Credential -> IO (Either String [Binding])
getBindingList did credential = do
  let path = "/api/v1/game/player/binding"
  signed <- signedHeaders credential did "GET" path ""
  case signed of
    Left err -> pure (Left err)
    Right signHeaders -> do
      response <- curlGet ("https://zonai.skland.com" <> path) (baseHeaders did <> [("cred", cred credential)] <> signHeaders)
      pure (response >>= parseBindings)

signArknights :: String -> Credential -> Binding -> IO GameSignResult
signArknights did credential binding = do
  let path = "/api/v1/game/attendance"
      body = arknightsBody binding
  signed <- signedHeaders credential did "POST" path body
  case signed of
    Left err -> pure (SignFailed "明日方舟" (Just (roleName binding)) err)
    Right signHeaders -> do
      response <- curlPost ("https://zonai.skland.com" <> path) (baseHeaders did <> [("cred", cred credential)] <> signHeaders) body
      pure (parseAttendanceResult "明日方舟" (roleName binding) response)

signEndfield :: String -> Credential -> Binding -> EndfieldRole -> IO GameSignResult
signEndfield did credential binding role = do
  let path = "/web/v1/game/endfield/attendance"
      body = ""
      roleHeader = "3_" <> roleId role <> "_" <> serverId role
      character =
        if null (roleNickname role)
          then roleName binding
          else roleNickname role
  signed <- signedHeaders credential did "POST" path body
  case signed of
    Left err -> pure (SignFailed "终末地" (Just character) err)
    Right signHeaders -> do
      response <-
        curlPost
          ("https://zonai.skland.com" <> path)
          ( baseHeaders did
              <> [ ("cred", cred credential)
                 , ("Content-Type", "application/json")
                 , ("sk-game-role", roleHeader)
                 , ("referer", "https://game.skland.com/")
                 , ("origin", "https://game.skland.com/")
                 ]
              <> signHeaders
          )
          body
      pure (parseAttendanceResult "终末地" character response)

curlGet :: String -> [(String, String)] -> IO (Either String String)
curlGet url headers = runCurl (["-sS", "--compressed", "-X", "GET", url] <> headerArgs headers)

curlPost :: String -> [(String, String)] -> String -> IO (Either String String)
curlPost url headers body =
  runCurl (["-sS", "--compressed", "-X", "POST", url] <> headerArgs headers <> ["--data-binary", body])

runCurl :: [String] -> IO (Either String String)
runCurl args = do
  (code, out, err) <- readCreateProcessWithExitCode (procSpec args) ""
  case code of
    ExitSuccess -> pure (Right out)
    ExitFailure _ -> pure (Left ("curl 失败: " <> err))

procSpec :: [String] -> CreateProcess
procSpec = proc "curl"

headerArgs :: [(String, String)] -> [String]
headerArgs headers = concat [["-H", key <> ": " <> value] | (key, value) <- headers]

parseAuthorization :: String -> Either String String
parseAuthorization raw = do
  json <- parseJson raw
  case lookupKey "status" json >>= asInt of
    Just 0 ->
      maybe (Left "OAuth 响应缺少 data.code") Right (lookupPath ["data", "code"] json >>= asString)
    statusValue ->
      Left ("OAuth 授权失败: " <> messageOf json <> " status=" <> show statusValue)

parseCredential :: String -> Either String Credential
parseCredential raw = do
  json <- parseJson raw
  case lookupKey "code" json >>= asInt of
    Just 0 -> do
      credentialCred <- maybe (Left "credential 响应缺少 data.cred") Right (lookupPath ["data", "cred"] json >>= asString)
      credentialToken <- maybe (Left "credential 响应缺少 data.token") Right (lookupPath ["data", "token"] json >>= asString)
      Right (Credential credentialCred credentialToken)
    codeValue ->
      Left ("credential 生成失败: " <> messageOf json <> " code=" <> show codeValue)

parseBindings :: String -> Either String [Binding]
parseBindings raw = do
  json <- parseJson raw
  case lookupKey "code" json >>= asInt of
    Just 0 -> do
      groups <- maybe (Left "绑定响应缺少 data.list") Right (lookupPath ["data", "list"] json >>= asArray)
      pure (concatMap parseGroup groups)
    codeValue ->
      Left ("查询绑定角色失败: " <> messageOf json <> " code=" <> show codeValue)

parseGroup :: Json -> [Binding]
parseGroup group =
  let code = stringField "appCode" group
      list = maybe [] id (lookupKey "bindingList" group >>= asArray)
   in map (parseBinding code) list

parseBinding :: String -> Json -> Binding
parseBinding code value =
  Binding
    { appCode = code
    , gameName = stringField "gameName" value
    , roleName = firstNonEmpty [stringField "nickName" value, stringField "nickname" value, stringField "name" value]
    , channelName = stringField "channelName" value
    , uid = stringField "uid" value
    , gameId = intField "gameId" value
    , roles = maybe [] (map parseRole) (lookupKey "roles" value >>= asArray)
    }

parseRole :: Json -> EndfieldRole
parseRole value =
  EndfieldRole
    { roleId = firstNonEmpty [stringField "roleId" value, stringField "id" value]
    , serverId = firstNonEmpty [stringField "serverId" value, stringField "server" value]
    , roleNickname = firstNonEmpty [stringField "nickname" value, stringField "nickName" value, stringField "name" value]
    }

parseAttendanceResult :: String -> String -> Either String String -> GameSignResult
parseAttendanceResult game character (Left err) = SignFailed game (Just character) err
parseAttendanceResult game character (Right raw) =
  case parseJson raw of
    Left err -> SignFailed game (Just character) ("签到响应 JSON 解析失败: " <> err)
    Right json ->
      case lookupKey "code" json >>= asInt of
        Just 0 -> SignSucceeded game character (parseAwards json)
        _ ->
          let msg = messageOf json
           in if isAlreadySigned msg
                then AlreadySigned game character msg
                else SignFailed game (Just character) msg

parseAwards :: Json -> [Award]
parseAwards json =
  case lookupPath ["data", "awards"] json >>= asArray of
    Just awards -> map parseArknightsAward awards
    Nothing -> parseEndfieldAwards json

parseArknightsAward :: Json -> Award
parseArknightsAward value =
  Award
    { awardName = firstNonEmpty [maybe "" id (lookupPath ["resource", "name"] value >>= asString), stringField "name" value]
    , awardCount = intField "count" value
    }

parseEndfieldAwards :: Json -> [Award]
parseEndfieldAwards json =
  let ids = maybe [] id (lookupPath ["data", "awardIds"] json >>= asArray)
      resourceMap = maybe [] id (lookupPath ["data", "resourceInfoMap"] json >>= asObject)
   in [ Award (resourceName awardId resourceMap) (resourceCount awardId resourceMap)
      | item <- ids
      , let awardId = firstNonEmpty [stringField "id" item, maybe "" id (asString item)]
      , not (null awardId)
      ]

resourceName :: String -> [(String, Json)] -> String
resourceName awardId resourceMap =
  case lookup awardId resourceMap of
    Just value -> firstNonEmpty [stringField "name" value, maybe "" id (lookupPath ["resource", "name"] value >>= asString), awardId]
    Nothing -> awardId

resourceCount :: String -> [(String, Json)] -> Int
resourceCount awardId resourceMap =
  case lookup awardId resourceMap of
    Just value -> max 1 (intField "count" value)
    Nothing -> 1

messageOf :: Json -> String
messageOf json = firstNonEmpty [maybe "" id (lookupKey "message" json >>= asString), maybe "" id (lookupKey "msg" json >>= asString), "未知错误"]

isAlreadySigned :: String -> Bool
isAlreadySigned msg = any (`contains` msg) ["已", "重复", "already", "signed"]

contains :: String -> String -> Bool
contains needle haystack = any (needle `prefixOf`) (tails haystack)

prefixOf :: String -> String -> Bool
prefixOf [] _ = True
prefixOf _ [] = False
prefixOf (x : xs) (y : ys) = x == y && prefixOf xs ys

tails :: [a] -> [[a]]
tails [] = [[]]
tails xs@(_ : rest) = xs : tails rest

firstNonEmpty :: [String] -> String
firstNonEmpty [] = ""
firstNonEmpty (x : xs)
  | null x = firstNonEmpty xs
  | otherwise = x

File: src/http.rs
use anyhow::{Context, Result, anyhow, bail};
use reqwest::header::{
    ACCEPT_ENCODING, CONNECTION, CONTENT_TYPE, HeaderMap, HeaderName, HeaderValue, ORIGIN, REFERER,
    USER_AGENT,
};
use reqwest::{Client, Method};
use serde_json::{Value, json};

use crate::sign::{arknights_body, signed_headers};
use crate::types::{Award, Binding, Credential, EndfieldRole, GameSignResult, UserConfig};

const USER_AGENT_VALUE: &str = "Mozilla/5.0 SKLand/1.52.1";
const AUTH_BASE_URL: &str = "https://as.hypergryph.com";
const ZONAI_BASE_URL: &str = "https://zonai.skland.com";

#[derive(Debug, Clone)]
pub struct SklandClient {
    client: Client,
}

impl SklandClient {
    pub fn new() -> Result<Self> {
        let client = Client::builder()
            .gzip(true)
            .build()
            .context("无法创建 HTTP client")?;
        Ok(Self { client })
    }

    pub async fn get_authorization(&self, did: &str, user: &UserConfig) -> Result<String> {
        let body = json!({
            "appCode": "4ca99fa6b56cc2ba",
            "token": user.user_token,
            "type": 0,
        })
        .to_string();
        let response = self
            .request(
                Method::POST,
                AUTH_BASE_URL,
                "/user/oauth2/v2/grant",
                base_headers(did),
                body,
            )
            .await?;
        parse_authorization(&response)
    }

    pub async fn get_credential(&self, did: &str, code: &str) -> Result<Credential> {
        let body = json!({
            "code": code,
            "kind": 1,
        })
        .to_string();
        let response = self
            .request(
                Method::POST,
                ZONAI_BASE_URL,
                "/web/v1/user/auth/generate_cred_by_code",
                base_headers(did),
                body,
            )
            .await?;
        parse_credential(&response)
    }

    pub async fn get_binding_list(
        &self,
        did: &str,
        credential: &Credential,
    ) -> Result<Vec<Binding>> {
        let path = "/api/v1/game/player/binding";
        let headers = signed_header_map(
            did,
            credential,
            &signed_headers(credential, did, "GET", path, "")?,
        )?;
        let response = self
            .request(Method::GET, ZONAI_BASE_URL, path, headers, String::new())
            .await?;
        parse_bindings(&response)
    }

    pub async fn sign_arknights(
        &self,
        did: &str,
        credential: &Credential,
        binding: &Binding,
    ) -> GameSignResult {
        let path = "/api/v1/game/attendance";
        let body = arknights_body(binding);
        let response = async {
            let headers = signed_header_map(
                did,
                credential,
                &signed_headers(credential, did, "POST", path, &body)?,
            )?;
            self.request(Method::POST, ZONAI_BASE_URL, path, headers, body)
                .await
        }
        .await;
        parse_attendance_result("明日方舟", &binding.role_name, response)
    }

    pub async fn sign_endfield(
        &self,
        did: &str,
        credential: &Credential,
        binding: &Binding,
        role: &EndfieldRole,
    ) -> GameSignResult {
        let path = "/web/v1/game/endfield/attendance";
        let role_header = format!("3_{}_{}", role.role_id, role.server_id);
        let character = if role.role_nickname.is_empty() {
            binding.role_name.as_str()
        } else {
            role.role_nickname.as_str()
        };
        let response = async {
            let mut headers = signed_header_map(
                did,
                credential,
                &signed_headers(credential, did, "POST", path, "")?,
            )?;
            headers.insert(
                HeaderName::from_static("sk-game-role"),
                HeaderValue::from_str(&role_header).context("非法 sk-game-role header")?,
            );
            headers.insert(
                REFERER,
                HeaderValue::from_static("https://game.skland.com/"),
            );
            headers.insert(ORIGIN, HeaderValue::from_static("https://game.skland.com/"));
            self.request(Method::POST, ZONAI_BASE_URL, path, headers, String::new())
                .await
        }
        .await;
        parse_attendance_result("终末地", character, response)
    }

    async fn request(
        &self,
        method: Method,
        base_url: &str,
        path: &str,
        headers: HeaderMap,
        body: String,
    ) -> Result<String> {
        let url = format!("{base_url}{path}");
        let mut request = self.client.request(method, url).headers(headers);
        if !body.is_empty() {
            request = request.body(body);
        }
        let response = request.send().await.context("HTTP 请求发送失败")?;
        let status = response.status();
        let text = response.text().await.context("HTTP 响应读取失败")?;
        if !status.is_success() {
            bail!("HTTP 请求失败: status={} body={}", status.as_u16(), text);
        }
        Ok(text)
    }
}

fn base_headers(did: &str) -> HeaderMap {
    let mut headers = HeaderMap::new();
    headers.insert(USER_AGENT, HeaderValue::from_static(USER_AGENT_VALUE));
    headers.insert(ACCEPT_ENCODING, HeaderValue::from_static("gzip"));
    headers.insert(CONNECTION, HeaderValue::from_static("close"));
    headers.insert(
        HeaderName::from_static("x-requested-with"),
        HeaderValue::from_static("com.hypergryph.skland"),
    );
    if let Ok(value) = HeaderValue::from_str(did) {
        headers.insert(HeaderName::from_static("did"), value);
    }
    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
    headers
}

fn signed_header_map(
    did: &str,
    credential: &Credential,
    signed: &crate::sign::SignedHeaders,
) -> Result<HeaderMap> {
    let mut headers = base_headers(did);
    headers.insert(
        HeaderName::from_static("cred"),
        HeaderValue::from_str(&credential.cred)?,
    );
    headers.insert(
        HeaderName::from_static("platform"),
        HeaderValue::from_str(&signed.platform)?,
    );
    headers.insert(
        HeaderName::from_static("timestamp"),
        HeaderValue::from_str(&signed.timestamp)?,
    );
    headers.insert(
        HeaderName::from_static("vname"),
        HeaderValue::from_str(&signed.version_name)?,
    );
    headers.insert(
        HeaderName::from_static("sign"),
        HeaderValue::from_str(&signed.sign)?,
    );
    Ok(headers)
}

fn parse_authorization(raw: &str) -> Result<String> {
    let json: Value = serde_json::from_str(raw).context("OAuth 响应 JSON 解析失败")?;
    match json.get("status").and_then(Value::as_i64) {
        Some(0) => json
            .pointer("/data/code")
            .and_then(Value::as_str)
            .map(ToOwned::to_owned)
            .ok_or_else(|| anyhow!("OAuth 响应缺少 data.code")),
        status => bail!("OAuth 授权失败: {} status={status:?}", message_of(&json)),
    }
}

fn parse_credential(raw: &str) -> Result<Credential> {
    let json: Value = serde_json::from_str(raw).context("credential 响应 JSON 解析失败")?;
    match json.get("code").and_then(Value::as_i64) {
        Some(0) => Ok(Credential {
            cred: json
                .pointer("/data/cred")
                .and_then(Value::as_str)
                .ok_or_else(|| anyhow!("credential 响应缺少 data.cred"))?
                .to_owned(),
            token: json
                .pointer("/data/token")
                .and_then(Value::as_str)
                .ok_or_else(|| anyhow!("credential 响应缺少 data.token"))?
                .to_owned(),
        }),
        code => bail!("credential 生成失败: {} code={code:?}", message_of(&json)),
    }
}

fn parse_bindings(raw: &str) -> Result<Vec<Binding>> {
    let json: Value = serde_json::from_str(raw).context("绑定响应 JSON 解析失败")?;
    match json.get("code").and_then(Value::as_i64) {
        Some(0) => {
            let groups = json
                .pointer("/data/list")
                .and_then(Value::as_array)
                .ok_or_else(|| anyhow!("绑定响应缺少 data.list"))?;
            Ok(groups.iter().flat_map(parse_group).collect())
        }
        code => bail!("查询绑定角色失败: {} code={code:?}", message_of(&json)),
    }
}

fn parse_group(group: &Value) -> Vec<Binding> {
    let app_code = string_field(group, "appCode");
    group
        .get("bindingList")
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .map(|item| parse_binding(&app_code, item))
                .collect()
        })
        .unwrap_or_default()
}

fn parse_binding(app_code: &str, value: &Value) -> Binding {
    Binding {
        app_code: app_code.to_owned(),
        game_name: string_field(value, "gameName"),
        role_name: first_non_empty(&[
            string_field(value, "nickName"),
            string_field(value, "nickname"),
            string_field(value, "name"),
        ]),
        channel_name: string_field(value, "channelName"),
        uid: string_field(value, "uid"),
        game_id: int_field(value, "gameId"),
        roles: value
            .get("roles")
            .and_then(Value::as_array)
            .map(|items| items.iter().map(parse_role).collect())
            .unwrap_or_default(),
    }
}

fn parse_role(value: &Value) -> EndfieldRole {
    EndfieldRole {
        role_id: first_non_empty(&[string_field(value, "roleId"), string_field(value, "id")]),
        server_id: first_non_empty(&[
            string_field(value, "serverId"),
            string_field(value, "server"),
        ]),
        role_nickname: first_non_empty(&[
            string_field(value, "nickname"),
            string_field(value, "nickName"),
            string_field(value, "name"),
        ]),
    }
}

fn parse_attendance_result(
    game: &str,
    character: &str,
    response: Result<String>,
) -> GameSignResult {
    match response {
        Err(err) => GameSignResult::SignFailed {
            game: game.to_owned(),
            character: Some(character.to_owned()),
            reason: err.to_string(),
        },
        Ok(raw) => match serde_json::from_str::<Value>(&raw) {
            Err(err) => GameSignResult::SignFailed {
                game: game.to_owned(),
                character: Some(character.to_owned()),
                reason: format!("签到响应 JSON 解析失败: {err}"),
            },
            Ok(json) => {
                if json.get("code").and_then(Value::as_i64) == Some(0) {
                    GameSignResult::SignSucceeded {
                        game: game.to_owned(),
                        character: character.to_owned(),
                        awards: parse_awards(&json),
                    }
                } else {
                    let message = message_of(&json);
                    if is_already_signed(&message) {
                        GameSignResult::AlreadySigned {
                            game: game.to_owned(),
                            character: character.to_owned(),
                            reason: message,
                        }
                    } else {
                        GameSignResult::SignFailed {
                            game: game.to_owned(),
                            character: Some(character.to_owned()),
                            reason: message,
                        }
                    }
                }
            }
        },
    }
}

fn parse_awards(json: &Value) -> Vec<Award> {
    if let Some(awards) = json.pointer("/data/awards").and_then(Value::as_array) {
        return awards.iter().map(parse_arknights_award).collect();
    }
    parse_endfield_awards(json)
}

fn parse_arknights_award(value: &Value) -> Award {
    Award {
        name: first_non_empty(&[
            value
                .pointer("/resource/name")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .to_owned(),
            string_field(value, "name"),
        ]),
        count: int_field(value, "count"),
    }
}

fn parse_endfield_awards(json: &Value) -> Vec<Award> {
    let Some(ids) = json.pointer("/data/awardIds").and_then(Value::as_array) else {
        return Vec::new();
    };
    let resource_map = json
        .pointer("/data/resourceInfoMap")
        .and_then(Value::as_object);
    ids.iter()
        .filter_map(|item| {
            let award_id = first_non_empty(&[
                string_field(item, "id"),
                item.as_str().unwrap_or_default().to_owned(),
            ]);
            if award_id.is_empty() {
                return None;
            }
            let resource = resource_map.and_then(|map| map.get(&award_id));
            Some(Award {
                name: resource_name(&award_id, resource),
                count: resource
                    .and_then(|value| value.get("count"))
                    .and_then(Value::as_i64)
                    .unwrap_or(1)
                    .max(1),
            })
        })
        .collect()
}

fn resource_name(award_id: &str, resource: Option<&Value>) -> String {
    resource.map_or_else(
        || award_id.to_owned(),
        |value| {
            first_non_empty(&[
                string_field(value, "name"),
                value
                    .pointer("/resource/name")
                    .and_then(Value::as_str)
                    .unwrap_or_default()
                    .to_owned(),
                award_id.to_owned(),
            ])
        },
    )
}

fn message_of(json: &Value) -> String {
    first_non_empty(&[
        json.get("message")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_owned(),
        json.get("msg")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_owned(),
        "未知错误".to_owned(),
    ])
}

fn is_already_signed(message: &str) -> bool {
    ["已", "重复", "already", "signed"]
        .iter()
        .any(|needle| message.contains(needle))
}

fn string_field(value: &Value, key: &str) -> String {
    value
        .get(key)
        .and_then(|value| {
            value
                .as_str()
                .map(ToOwned::to_owned)
                .or_else(|| value.as_i64().map(|n| n.to_string()))
        })
        .unwrap_or_default()
}

fn int_field(value: &Value, key: &str) -> i64 {
    value
        .get(key)
        .and_then(|value| {
            value
                .as_i64()
                .or_else(|| value.as_str().and_then(|value| value.parse().ok()))
        })
        .unwrap_or(0)
}

fn first_non_empty(values: &[String]) -> String {
    values
        .iter()
        .find(|value| !value.is_empty())
        .cloned()
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_binding_groups() {
        let raw = r#"{"code":0,"data":{"list":[{"appCode":"arknights","bindingList":[{"gameName":"明日方舟","nickName":"博士","uid":"1","gameId":1}]}]}}"#;

        let bindings = parse_bindings(raw).unwrap();

        assert_eq!(bindings.len(), 1);
        assert_eq!(bindings[0].app_code, "arknights");
        assert_eq!(bindings[0].role_name, "博士");
    }
}
